src/Native/Runtime/amd64/AllocFast.asm

```diff
@@ -101,6 +101,56 @@ NewOutOfMemory:
         jmp         RhExceptionHandling_FailedAllocation
 NESTED_END RhpNewObject, _TEXT
 
+;; Allocate one dimensional, zero based array (SZARRAY).
+;;  RCX == EEType
+;;  EDX == element count
+LEAF_ENTRY RhNewString, _TEXT
+
+        ; we want to limit the element count to the non-negative 32-bit int range
+        cmp         rdx, 07fffffffh
+        ja          StringSizeOverflow
+
+        ; Compute overall allocation size (align(base size + (element size * elements), 8)).
+        lea         rax, [rdx * 2 + (2 * 8 + 4 + 2) + 7]
+        and         rax, -8
+
+        ; rax == array size
+        ; rcx == EEType
+        ; rdx == element count
+
+        INLINE_GETTHREAD r10, r8
+
+        mov         r8, rax
+        add         rax, [r10 + OFFSETOF__Thread__m_alloc_context__alloc_ptr]
+        jc          RhpNewArrayRare
```

Comment:
The overflow case is indirectly covered by this check in CoreCLR: https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/vm/amd64/JitHelpers_InlineGetThread.asm#L170. The CoreCLR implementation depends on having at least `LARGE_OBJECT_SIZE` reserved at the end of the address space.